osmdroid/osmdroid
# Feature/#906 ESRI ShapeFile reader + Bounds checks on drawing lines, polygons and markers

New module, ESRI Shape Reader. This converts shape file contents into polygons with the associated metadata attached to the infowindow popup. This is actually the smallest of the changes on this branch.

The large changes is optimizations to Polyline, Polygon, Marker, and Itemized Icon Overlay. As suggested by @MKergall in #770, this uses the bounding box check for those mentioned classes in the on draw method. If there is no overlap of the bounds of the overlay with the view's bounds, then skip the draw cycle.

Finally, assuming the line or polygon is within bounds, the point reduction algorithm is ran before drawing. At numerically low zoom levels, this usually reduces small scale polygons down to a triangle and it becomes more and more accurate the further in your zoom.

Impacts:
 - Bounds check: At numerically high zoom levels with a hundreds of geographically dispersed polygons, this has a huge performance boost (HUGE).
 - Point reduction: This has a noticeable impact at numerically low zoom levels
 - Polygons now have a "minimum zoom" level setting to prevent drawing a very low zoom levels. Default is 0 of course but it can be set to bypass drawing if needed for certain applications

Current open issues
 - Bounds check with info window. When the item scrolls out of view, the info window appears to cling to the edge of the screen. I did attempt to auto hide it but ran into some issues with it. I'm not sure what "right" looks like here.
 - When looking at the world at zoom <= 3, the bounds checks are skipped entirely due to world replication issues and no way to check if there are multiple worlds or not.

How to test
 - see #906, there's a download link for a shape file for country borders. Unzip it and copy the contents of the zip to somewhere on your device.
 - Open the app, more samples, Data, Shape File Import, hit the ... button, "Import a shape file"
 - then select the .shp file from the zip contents
